inspirehep/modules/orcid/cache.py

```diff
@@ -31,69 +31,114 @@
+        self._new_hash_value = None
+        if not _redis_client:
+            _init_redis_client()
+        self.redis = _redis_client
 
-    def _get_key(self, recid):
+    def get_key(self, recid):
         """Return the string 'orcidcache:``orcid_value``:``recid``'"""
         return 'orcidcache:{}:{}'.format(self.orcid, recid)
 
-    def write_record_data(self, recid, putcode, hash_value=None):
+    def write_work_putcode(self, recid, putcode, inspire_record=None):
         """
         Write the putcode and the hash for the given (orcid, recid).
 
         Args:
-            recid(int): inspire record's id.
-            putcode(string): the putcode used to push the record to ORCID.
-            hash_value(Optional[string]): hashed ORCID record content.
+            recid (string): record identifier.
+            putcode (string): the putcode used to push the record to ORCID.
+            inspire_record (InspireRecord): InspireRecord instance. If provided,
+             the hash for the record content is re-computed.
         """
         if not putcode:
             raise EmptyPutcodeError
 
-        key = self._get_key(recid)
-        value = {'putcode': putcode}
-        if hash_value:
-            value['hash'] = hash_value
-        self.redis.hmset(key, value)
+        data = {'putcode': putcode}
+
+        if inspire_record:
+            if not self._new_hash_value:
+                self._new_hash_value = _OrcidHasher(inspire_record).compute_hash()
+            data['hash'] = self._new_hash_value
 
-    def read_record_data(self, recid):
+        key = self.get_key(recid)
+        self.redis.hmset(key, data)
+
+    def read_work_putcode(self, recid):
+        """Read the putcode for the given (orcid, recid)."""
+        key = self.get_key(recid)
+        value = self.redis.hgetall(key)
+        self._cached_hash_value = value.get('hash')
+        return value.get('putcode')
+
+    def has_work_content_changed(self, recid, inspire_record):
         """
-        Read the putcode and the hash for the given (orcid, recid).
+        True if the work content has changed compared to the cached version.
 
         Args:
-            recid(int): inspire record's id.
+            recid (string): record identifier.
+            inspire_record (InspireRecord): InspireRecord instance. If provided,
+             the hash for the record content is re-computed.
         """
-        key = self._get_key(recid)
-        value = self.redis.hgetall(key)
-        return value.get('putcode'), value.get('hash')
+        if not self._cached_hash_value:
+            self.read_work_putcode(recid)
+        if not self._new_hash_value:
+            self._new_hash_value = _OrcidHasher(inspire_record).compute_hash()
+        return self._cached_hash_value != self._new_hash_value
 
 
-class OrcidHasher(object):
-    def compute_hash(self, record):
-        """Generate hash for an ORCID-serialised HEP record
+class _OrcidHasher(object):
```

Comment:
I know we've already discussed this, but a class with an `__init__` and one public method is not a class, it's a function. In the meantime I remembered this good talk https://www.youtube.com/watch?reload=9&v=o9pEzgHorH0 with a more detailed explanation.

Reply:
I explained the reasons to group 3 related functions in a single object. Class/static methods are pretty popular all over, even in CPython lib source code

Reply:
I can add another thing: our codebase is a jungle of code without architecture. Grouping code into abstractions that goes beyond the `utils` module is what we need.
A class goes into this direction, even in this case which I recognize being not a perfect use case.

As for the over use of classes, I recognize the problems coming with over use of inheritance. For the rest in my experience, OOP was always beneficial to code architecture

Reply:
Also, I watched the video and he talks about this use case in the last question: https://youtu.be/o9pEzgHorH0?t=1594

Reply:
I don't think that question is exactly your use case, but I agree that anything that goes in the direction of better code organization is positive. So let's not waste more time on bikeshedding and agree to disagree.